README.markdown
# PyOCR

PyOCR is an optical character recognition (OCR) tool wrapper for python.
That is, it helps using various OCR tools from a Python program.

It has been tested only on GNU/Linux systems. It should also work on similar
systems (*BSD, etc). It may or may not work on Windows, MacOSX, etc.


## Supported OCR tools

* Libtesseract (Python bindings for the C API)
* Tesseract (wrapper: fork + exec)
* Cuneiform (wrapper: fork + exec)

## Features

* Supports all the image formats supported by [Pillow](https://github.com/python-imaging/Pillow),
  including jpeg, png, gif, bmp, tiff and others
* Various output types: text only, bounding boxes, etc.
* Orientation detection (Tesseract and libtesseract only)
* Can focus on digits only (Tesseract and libtesseract only)
* Can save and reload boxes in hOCR format
* PDF generation (libtesseract only)


## Limitations

* hOCR: Only a subset of the specification is supported. For instance, pages and
  paragraph positions are not stored.


## Installation

```sh
sudo pip install pyocr  # Python 2.7
sudo pip3 install pyocr  # Python 3.X
```

or the manual way:
```sh
mkdir -p ~/git ; cd git
git clone https://github.com/openpaperwork/pyocr.git
cd pyocr
make install  # will run 'python ./setup.py install'
```


## Usage

### Initialization

```Python
from PIL import Image
import sys

import pyocr
import pyocr.builders

tools = pyocr.get_available_tools()
if len(tools) == 0:
    print("No OCR tool found")
    sys.exit(1)
# The tools are returned in the recommended order of usage
tool = tools[0]
print("Will use tool '%s'" % (tool.get_name()))
# Ex: Will use tool 'libtesseract'

langs = tool.get_available_languages()
print("Available languages: %s" % ", ".join(langs))
lang = langs[0]
print("Will use lang '%s'" % (lang))
# Ex: Will use lang 'fra'
# Note that languages are NOT sorted in any way. Please refer
# to the system locale settings for the default language
# to use.
```

### Image to text

```Python
txt = tool.image_to_string(
    Image.open('test.png'),
    lang=lang,
    builder=pyocr.builders.TextBuilder()
)
# txt is a Python string

word_boxes = tool.image_to_string(
    Image.open('test.png'),
    lang="eng",
    builder=pyocr.builders.WordBoxBuilder()
)
# list of box objects. For each box object:
#   box.content is the word in the box
#   box.position is its position on the page (in pixels)
#
# Beware that some OCR tools (Tesseract for instance)
# may return empty boxes

line_and_word_boxes = tool.image_to_string(
    Image.open('test.png'), lang="fra",
    builder=pyocr.builders.LineBoxBuilder()
)
# list of line objects. For each line object:
#   line.word_boxes is a list of word boxes (the individual words in the line)
#   line.content is the whole text of the line
#   line.position is the position of the whole line on the page (in pixels)
#
# Each word box object has an attribute 'confidence' giving the confidence
# score provided by the OCR tool. Confidence score depends entirely on
# the OCR tool. Only supported with Tesseract and Libtesseract (always 0
# with Cuneiform).
#
# Beware that some OCR tools (Tesseract for instance) may return boxes
# with an empty content.

# Digits - Only Tesseract (not 'libtesseract' yet !)
digits = tool.image_to_string(
    Image.open('test-digits.png'),
    lang=lang,
    builder=pyocr.tesseract.DigitBuilder()
)
# digits is a python string
```

Argument 'lang' is optional. The default value depends of
the tool used.

Argument 'builder' is optional. Default value is
builders.TextBuilder().

If the OCR fails, an exception ```pyocr.PyocrException```
will be raised.

An exception MAY be raised if the input image contains no
text at all (depends on the OCR tool behavior).


### Orientation detection

Currently only available with Tesseract or Libtesseract.

```Python
if tool.can_detect_orientation():
    try:
        orientation = tool.detect_orientation(
            Image.open('test.png'),
            lang='fra'
        )
    except pyocr.PyocrException as exc:
        print("Orientation detection failed: {}".format(exc))
        return
    print("Orientation: {}".format(orientation))
# Ex: Orientation: {
#   'angle': 90,
#   'confidence': 123.4,
# }
```

Angles are given in degrees (range: [0-360[). Exact possible
values depend of the tool used. Tesseract only returns angles =
0, 90, 180, 270.

Confidence is a score arbitrarily defined by the tool. It MAY not
be returned.

detect_orientation() MAY raise an exception if there is no text
detected in the image.


### Writing and reading text files

Writing:

```Python
import codecs
import pyocr
import pyocr.builders

tool = pyocr.get_available_tools()[0]
builder = pyocr.builders.TextBuilder()

txt = tool.image_to_string(
    Image.open('test.png'),
    lang=lang,
    builder=builder
)
# txt is a Python string

with codecs.open("toto.txt", 'w', encoding='utf-8') as file_descriptor:
    builder.write_file(file_descriptor, txt)
# toto.txt is a simple text file, encoded in utf-8
```

Reading:

```Python
import codecs
import pyocr.builders

builder = pyocr.builders.TextBuilder()
with codecs.open("toto.txt", 'r', encoding='utf-8') as file_descriptor:
    txt = builder.read_file(file_descriptor)
# txt is a Python string
```

### Writing and reading hOCR files

Writing:

```Python
import codecs
import pyocr
import pyocr.builders

tool = pyocr.get_available_tools()[0]
builder = pyocr.builders.LineBoxBuilder()

line_boxes = tool.image_to_string(
    Image.open('test.png'),
    lang=lang,
    builder=builder
)
# list of LineBox (each box points to a list of word boxes)

with codecs.open("toto.html", 'w', encoding='utf-8') as file_descriptor:
    builder.write_file(file_descriptor, line_boxes)
# toto.html is a valid XHTML file
```

Reading:

```Python
import codecs
import pyocr.builders

builder = pyocr.builders.LineBoxBuilder()
with codecs.open("toto.html", 'r', encoding='utf-8') as file_descriptor:
    line_boxes = builder.read_file(file_descriptor)
# list of LineBox (each box points to a list of word boxes)
```


### Generating PDF file from an image

With libtesseract >= 4, it's possible to generate a PDF from an image:

```Python
import PIL.Image
import pyocr

pyocr.libtesseract.image_to_pdf(
    PIL.Image.open("image.jpg"),
    "output_filename"  # .pdf will be appended
)

```

Beware this code hasn't been adapted to libtesseract 3 yet.


## Dependencies

* PyOCR requires python 2.7 or later. Python 3 is supported.
* You will need [Pillow](https://github.com/python-imaging/Pillow)
  or Python Imaging Library (PIL). Under Debian/Ubuntu, Pillow is in
  the package ```python-pil``` (```python3-pil``` for the Python 3
  version).
* Install an OCR:
  * [libtesseract](http://code.google.com/p/tesseract-ocr/)
    ('libtesseract3' + 'tesseract-ocr-&lt;lang&gt;' in Debian).
  * or [tesseract-ocr](http://code.google.com/p/tesseract-ocr/)
    ('tesseract-ocr' + 'tesseract-ocr-&lt;lang&gt;' in Debian).
    You must be able to invoke the tesseract command as "tesseract".
    PyOCR is tested with Tesseract >= 3.01 only.
  * or Cuneiform


## Tests

```sh
make check  # requires pyflake8
make test  # requires tox, pytest and python3
```

Tests are made to be run without external dependencies (no Tesseract or Cuneiform needed).


## OCR on natural scenes

If you want to run OCR on natural scenes (photos, etc), you will have to filter
the image first. There are many algorithms possible to do that. One of those
who gives the best results is
[Stroke Width Transform](https://github.com/openpaperwork/libpillowfight#stroke-width-transformation).


## Contact

* [Mailing-list](https://github.com/openpaperwork/paperwork/wiki/Contact#mailing-list)
* [Bug tracker](https://github.com/openpaperwork/pyocr/issues/)


## Applications that use PyOCR

* [Mayan EDMS](http://mayan-edms.com/)
* [Paperless](https://github.com/danielquinn/paperless#readme)
* [Paperwork](https://github.com/openpaperwork/paperwork#readme)

If you know of any other applications that use Pyocr, please
[tell us](https://github.com/openpaperwork/paperwork/wiki/Contact#mailing-list) :-)

## Copyright

PyOCR is released under the GPL v3+.
Copyright belongs to the authors of each piece of code
(see the file AUTHORS for the contributors list, and
```git blame``` to know which lines belong to which author).

https://github.com/openpaperwork/pyocr


ChangeLog
18/02/2019 - 0.6:
- Complete rewrite of unit tests (thanks to Thomas Perret)
- Libtesseract 4.0: Fix segfault when running orientation detection
  (thanks to Marián Skrip)
- Libtesseract 4.0: Add a workaround: Tesseract need the locale to be set to
  'C' (thanks to Thomas Perret)
- Libtesseract: Specify DPI of the image to Libtesseract (thanks to Thomas
  Perret)
- Tesseract 4.0: Improve Tesseract version parsing

09/04/2018 - 0.5.3:
- Really fix tesseract 4.0 support (thanks to David Martin)
- Tests: switch from nose to pytest (thanks to Elliott Sales de Andrade)

25/07/2018 - 0.5.2:
- Fix tesseract 4.0 support: Use option '--psm' instead of '-psm'
- tesseract.detection_orientation(): Fix exception generation

01/03/2017 - 0.5.1:
- libtesseract/Windows: Add possible DLL names for libtesseract
- libtesseract: Keep track of library-loading errors in
  pyocr.libtesseract.lib_load_errors (useful for debugging)
- Build method has been changed: Use now "make install" instead of
  "python3 ./setup.py install"
- cosmetic: builders/WordHTMLParser: Message "OCR confidence not found"
  floods the logs when working with old documents --> switch to debug
  instead of info.

14/12/2017 - 0.5:
- Tesseract/Libtesseract + LineBoxBuilder: Add confidence scores to
  every word boxes and to hOCR files (thanks to Adriano Pagano)
- Tesseract 4 (shell): Add '--oem 0' to specify legacy model when doing
  orientation detection as orientation detection does not work yet with
  Tesseract 4 (thanks to Adriano Pagano)
- Libtesseract: Fix multi-language support
- Tesseract (shell) + Windows: Never let the cmd window appear
- Libtesseract: Implements image_to_pdf() (thanks to Marian Skrip)
- Libtesseract: Hide debug messages (thanks to Ashish Kulkarni)

13/05/2017 - 0.4.7:
- Tesseract 4.00.00alpha:
  - Version parsing: Ignore suffix (so '4.00.00alpha' == (4, 0, 0))
  - Libtesseract: Load libtesseract.so.4 instead of libtesseract.so.3 if
    available
- Support for Tesseract 3.05.00:
  - Builders: Split field 'tess_conf' into 'tess_flags' and 'tess_conf'
  - Libtesseract: If available, use TessBaseAPIDetectOrientationScript()
    instead of TessBaseAPIDetectOS
- Libtesseract: Workaround: Prevents possible segfault in image_to_string()
  when the target language is not available

26/01/2017 - 0.4.6:
- hOCR outputs: Generate valid XHTML files

10/01/2017 - 0.4.5:
- Clean up exceptions raised when OCR fails:
  - Now, all tools raise only exceptions inheriting from
    pyocr.PyocrException
  - There is now one and only one TesseractError (shared between
    pyocr.libtesseract and pyocr.tesseract)

08/12/2016 - 0.4.4:
- Fix Python 2.7 support (broken import)

06/12/2016 - 0.4.3:
- (temporary) Use tesseract-sh by default instead of libtesseract. Some
  people have reported crashes with Paperwork+libtesseract. It needs more
  stress-testing
- DigitBuilder is now available in 'pyocr.builders' (can be used
  with libtesseract and cuneiform)
- New builder: DigitLineBoxBuilder
- Windows: Fix pyinstaller packaging suport: env variable TESSDATA_PREFIX
  wasn't set correctly
- Windows: Tesseract-sh: Prevent CMD windows from appearing

05/10/2016 - 0.4.2:
* Tesseract: orientation detection: Ignore errors printed by libleptonic
  on stderr (thanks to TeisD)
* Tesseract: Fix support of dev builds (thanks to Fjup)
* Libtesseract: Fix support of dev builds (thanks to Jakub Semerák)
* Tesseract: Use '--list-langs' to get the available languages instead of
  looking for the data directory (thanks to Bernhard Liebl)

06/04/2016 - 0.4.1:
* Disable 'libtesseract' with Tesseract <= 3.03. It tends to segfault.
  Libtesseract: Disable it with Tesseract <= 3.03. It tends to segfault.
  Note: the segfault may not actually be related to Libtesseract. It may be due to other things in Debian stable (jessie).
  Anyway, Paperwork cannot work on Debian stable because of that --> disabled just to be safe


13/03/2013 - 0.4.0:
* New module: 'libtesseract'. Use the C API of Tesseract for OCR.
  This module is more efficient and cleaner than the old 'tesseract' module
  (no more fork + exec + sh, less image manipulation, etc).
  Note that with this module the images are just loaded and uncompressed
  by Pillow. With 'tesseract', they were loaded, uncompressed, re-compressed
  and saved by Pillow, then be reloaded by Leptonica. So the results may
  vary slightly.
* Tesseract: Add support for Win32
* Tesseract: Fix orientation detection for version >= 3.04.01


0.3.1:
* tesseract.detect_orientation(): Use a temporary file instead of stdin
  to transmit the image to Tesseract. Tesseract 3.04 doesn't support
  stdin + "-psm 0" (regression ?)
* tesseract.detect_orientation(): Improve output parsing reliability
* optim: Avoid unnecessary convert to RGB and allow using image formats
  different from PNG
* TextBuilder + Cuneiform: add extra settings for Cuneiform
  (cuneiform_dotmatrix, cuneiform_fax=False, cuneiform_singlecolumn)


0.3.0:
* New API: pyocr.<tool>.can_detect_orientation() and
  pyocr.<tool>.detect_orientation()


0.2.4:
* Tesseract : add digit-only support
* Tesseract : add support for Tesseract subsets of layout analysis (-psm)


0.2.3:
* Strip the alpha channel from images before running the OCR. It's basically
  useless and can prevent the tool from working correctly.
* Make hOCR parsing more resistant (handle extra data around box positions)
* Fix: Take into account that new versions of Tesseract uses the file
  extension .hocr instead of .html


0.2.2:
* Fix Python 3 support
* Add support for Tesseract on Heroku


0.2.1:
* Make it possible to use 'import pyocr' instead of 'from pyocr import pyocr'.
  'from pyocr import pyocr' still works but is obsolete.
* Fix dependency list: depends on Pillow (it's untested with PIL)
* Fix pyocr.VERSION


0.2.0:
* Python 3.x support


0.1.2:
* Tesseract: Fix version parsing
* Tesseract: Fix Tesseract 3.02.01's hOCR format support


0.1.1:
* hOCR: Parse lines as well as words

src/pyocr/tesseract.py
#!/usr/bin/env python
'''
tesseract.py is a wrapper for google's Tesseract-OCR
( http://code.google.com/p/tesseract-ocr/ ).

USAGE:
 > from PIL import Image
 > from pyocr.tesseract import image_to_string
 > print(image_to_string(Image.open('test.png')))
 > print(image_to_string(Image.open('test-european.jpg'), lang='fra'))

COPYRIGHT:
PyOCR is released under the GPL v3.
Copyright (c) Samuel Hoffstaetter, 2009
Copyright (c) Jerome Flesch, 2011-2016
https://github.com/openpaperwork/pyocr#readme
'''

import codecs
import errno
import logging
import os
import subprocess
import sys
import tempfile
import contextlib
import shutil

from . import builders
from . import util
from .builders import DigitBuilder  # backward compatibility
from .error import TesseractError  # backward compatibility
from .util import digits_only

try:
    FileNotFoundError
except NameError:
    # python2 does not have FileNotFoundError
    FileNotFoundError = IOError

# CHANGE THIS IF TESSERACT IS NOT IN YOUR PATH, OR IS NAMED DIFFERENTLY
TESSERACT_CMD = 'tesseract.exe' if os.name == 'nt' else 'tesseract'

TESSDATA_EXTENSION = ".traineddata"

logger = logging.getLogger(__name__)

g_subprocess_startup_info = None
g_creation_flags = 0

__all__ = [
    'CharBoxBuilder',
    'DigitBuilder',
    'can_detect_orientation',
    'detect_orientation',
    'get_available_builders',
    'get_available_languages',
    'get_name',
    'get_version',
    'image_to_string',
    'is_available',
    'TesseractError',
]


class CharBoxBuilder(builders.BaseBuilder):
    """
    If passed to image_to_string(), image_to_string() will return an array of
    Box. Each box correspond to a character recognized in the image.
    """

    def __init__(self):
        file_ext = ["box"]
        tess_flags = []
        tess_conf = ["batch.nochop", "makebox"]
        cun_args = []
        super(CharBoxBuilder, self).__init__(file_ext, tess_flags, tess_conf,
                                             cun_args)
        self.tesseract_layout = 1

    @staticmethod
    def read_file(file_descriptor):
        """
        Extract of set of Box from the lines of 'file_descriptor'

        Return:
            An array of Box.
        """
        boxes = []  # note that the order of the boxes may matter to the caller
        for line in file_descriptor.readlines():
            line = line.strip()
            if line == "":
                continue
            elements = line.split(" ")
            if len(elements) < 6:
                continue
            position = ((int(elements[1]), int(elements[2])),
                        (int(elements[3]), int(elements[4])))
            box = builders.Box(elements[0], position)
            boxes.append(box)
        return boxes

    @staticmethod
    def write_file(file_descriptor, boxes):
        """
        Write boxes in a box file. Output is in a the same format than
        tesseract's one.

        Warning:
            The file_descriptor must support UTF-8 ! (see module 'codecs')
        """
        for box in boxes:
            file_descriptor.write(box.get_unicode_string() + " 0\n")

    def __str__(self):
        return "Character boxes"


def _set_environment():
    global g_subprocess_startup_info
    global g_creation_flags

    if os.name == "nt":  # pragma: no cover
        g_subprocess_startup_info = subprocess.STARTUPINFO()
        g_subprocess_startup_info.wShowWindow = subprocess.SW_HIDE
        g_subprocess_startup_info.dwFlags |= subprocess.STARTF_USESHOWWINDOW
        g_creation_flags = 0x08000000  # CREATE_NO_WINDOW

    if getattr(sys, 'frozen', False):  # pragma: no cover
        # Pyinstaller support
        path = os.environ["PATH"]
        if sys._MEIPASS in path:
            # already changed
            return

        tesspath = os.path.join(sys._MEIPASS, "tesseract")
        tessprefix = os.path.join(sys._MEIPASS, "data")
        logger.info("Running in packaged environment")

        if not os.path.exists(os.path.join(tessprefix, "tessdata")):
            logger.warning(
                "Running from container, but no tessdata ({}) found !".format(
                    tessprefix
                )
            )
        else:
            logger.info("TESSDATA_PREFIX set to [{}]".format(tessprefix))
            os.environ['TESSDATA_PREFIX'] = tessprefix
        if not os.path.exists(tesspath):
            logger.warning(
                "Running from container, but no tesseract ({}) found !".format(
                    tesspath
                )
            )
        else:
            logger.info("[{}] added to PATH".format(tesspath))
            os.environ['PATH'] = (
                tesspath + os.pathsep + os.environ['PATH']
            )


def can_detect_orientation():
    version = get_version()
    return (
        version[0] > 3 or
        (version[0] == 3 and version[1] >= 3)
    )


def psm_parameter():
    """Return the psm option string depending on the Tesseract version."""
    version = get_version()
    return "--psm" if version[0] > 3 else "-psm"


def detect_orientation(image, lang=None):
    """
    Arguments:
        image --- Pillow image to analyze
        lang --- lang to specify to tesseract

    Returns:
        {
            'angle': 90,
            'confidence': 23.73,
        }

    Raises:
        TesseractError --- if no script detected on the image
    """
    _set_environment()
    with temp_dir() as tmpdir:
        command = [TESSERACT_CMD, "input.bmp", 'stdout', psm_parameter(), "0"]
        version = get_version()
        if lang is not None:
            if version[0] < 4:
                command += ['-l', lang]
            else:
                command += ['-l', 'osd']

        if image.mode != "RGB":
            image = image.convert("RGB")
        image.save(os.path.join(tmpdir, "input.bmp"))

        proc = subprocess.Popen(command, stdin=subprocess.PIPE, shell=False,
                                startupinfo=g_subprocess_startup_info,
                                creationflags=g_creation_flags,
                                cwd=tmpdir,
                                stdout=subprocess.PIPE,
                                stderr=subprocess.STDOUT)
        proc.stdin.close()
        original_output = proc.stdout.read()
        proc.wait()

        original_output = original_output.decode("utf-8")
        original_output = original_output.strip()

        if "Could not initialize tesseract" in original_output:
            raise TesseractError(-1, "Error initializing tesseract: %s"
                                 % original_output)

        try:
            output = original_output.split("\n")
            output = [line.split(": ", 1) for line in output if (": " in line)]
            output = {x: y for (x, y) in output}
            angle = int(output.get('Rotate', output['Orientation in degrees']))
            # Tesseract reports the angle in the opposite direction the one we
            # want
            angle = (360 - angle) % 360
            return {
                'angle': angle,
                'confidence': float(output['Orientation confidence']),
            }
        except Exception as ex:
            raise TesseractError(-1, "No script found in image (%s - %s)"
                                 % (str(ex), original_output))


def get_name():
    return "Tesseract (sh)"


def get_available_builders():
    return [
        builders.LineBoxBuilder,
        builders.TextBuilder,
        builders.WordBoxBuilder,
        CharBoxBuilder,
        builders.DigitBuilder,
        builders.DigitLineBoxBuilder,
    ]


def run_tesseract(input_filename, output_filename_base, cwd=None, lang=None,
                  flags=None, configs=None):
    '''
    Runs Tesseract:
        `TESSERACT_CMD` \
                `input_filename` \
                `output_filename_base` \
                [-l `lang`] \
                [`configs`]

    Arguments:
        input_filename --- image to read
        output_filename_base --- file name in which must be stored the result
            (without the extension)
        cwd --- Run Tesseract in the specified working directory or use current
            one if None
        lang --- Tesseract language to use (if None, none will be specified)
        config --- List of Tesseract configs to use (if None, none will be
            specified)

    Returns:
        Returns (the exit status of Tesseract, Tesseract's output)
    '''
    _set_environment()

    command = [TESSERACT_CMD, input_filename, output_filename_base]

    if lang is not None:
        command += ['-l', lang]

    if flags is not None:
        command += flags

    if configs is not None:
        command += configs

    proc = subprocess.Popen(command, cwd=cwd,
                            startupinfo=g_subprocess_startup_info,
                            creationflags=g_creation_flags,
                            stdout=subprocess.PIPE,
                            stderr=subprocess.STDOUT)
    # Beware that in some cases, tesseract may print more on stderr than
    # allowed by the buffer of subprocess.Popen.stderr. So we must read stderr
    # asap or Tesseract will remain stuck when trying to write again on stderr.
    # In the end, we just have to make sure that proc.stderr.read() is called
    # before proc.wait()
    errors = proc.stdout.read()
    return (proc.wait(), errors)


def cleanup(filename):
    ''' Tries to remove the given filename. Ignores non-existent files '''
    try:
        os.remove(filename)
    except OSError:  # pragma: no cover
        pass


class ReOpenableTempfile(object):  # pragma: no cover
    """
    On Windows, `tempfile.NamedTemporaryFile` occur Permission denied Error
    when file is still open.
    It returns `tempfile.NamedTemporaryFile` compatible object.
    """
    def __init__(self, suffix):
        self.name = None
        with tempfile.NamedTemporaryFile(prefix='tess_', suffix=suffix,
                                         delete=False) as fp:
            self.name = fp.name

    def __enter__(self):
        return self

    def __exit__(self, type, value, traceback):
        self.close()

    def close(self):
        if self.name is not None:
            os.remove(self.name)
            self.name = None


@contextlib.contextmanager
def temp_dir():
    """
    A context manager for maintaining a temporary directory
    """
    # NOTE: Drop this as soon as we don't support Python 2.7 anymore, because
    # since Python 3.2 there is a context manager called TemporaryDirectory().
    path = tempfile.mkdtemp(prefix='tess_')
    try:
        yield path
    finally:
        shutil.rmtree(path)


def image_to_string(image, lang=None, builder=None):
    '''
    Runs tesseract on the specified image. First, the image is written to disk,
    and then the tesseract command is run on the image. Tesseract's result is
    read, and the temporary files are erased.

    Arguments:
        image --- image to OCR.
        lang --- tesseract language to use.
        builder --- builder used to configure Tesseract and read its result.
            The builder is used to specify the type of output expected.
            Possible builders are TextBuilder or CharBoxBuilder. If builder ==
            None, the builder used will be TextBuilder.

    Returns:
        Depends of the specified builder. By default, it will return a simple
        string.
    '''

    if builder is None:
        builder = builders.TextBuilder()
    with temp_dir() as tmpdir:
        if image.mode != "RGB":
            image = image.convert("RGB")
        image.save(os.path.join(tmpdir, "input.bmp"))
        (status, errors) = run_tesseract("input.bmp", "output", cwd=tmpdir,
                                         lang=lang,
                                         flags=builder.tesseract_flags,
                                         configs=builder.tesseract_configs)
        if status:
            raise TesseractError(status, errors)

        tested_files = []
        output_file_name = "ERROR"
        for file_extension in builder.file_extensions:
            output_file_name = ('%s.%s' % (os.path.join(tmpdir, "output"),
                                           file_extension))

            tested_files.append(output_file_name)
            try:
                with codecs.open(output_file_name, 'r', encoding='utf-8',
                                 errors='replace') as file_desc:
                    return builder.read_file(file_desc)
            except FileNotFoundError as exc:
                if sys.version_info < (3, 0):
                    # python2 has no FileNotFoundError specifid Exception
                    # so we rely on the errno of the IOError exception
                    if exc.errno == errno.ENOENT:
                        # file not found
                        continue
                    else:
                        raise exc
                continue
            finally:
                cleanup(output_file_name)
        raise TesseractError(
            -1, "Unable to find output file (tested {})".format(tested_files)
        )


def is_available():
    _set_environment()
    return util.is_on_path(TESSERACT_CMD)


def get_available_languages():
    """
    Returns the list of languages that Tesseract knows how to handle.

    Returns:
        An array of strings. Note that most languages name conform to ISO 639
        terminology, but not all. Most of the time, truncating the language
        name name returned by this function to 3 letters should do the trick.
    """
    _set_environment()
    proc = subprocess.Popen([TESSERACT_CMD, "--list-langs"],
                            startupinfo=g_subprocess_startup_info,
                            creationflags=g_creation_flags,
                            stdout=subprocess.PIPE,
                            stderr=subprocess.STDOUT)
    langs = proc.stdout.read().decode('utf-8').splitlines(False)
    ret = proc.wait()
    if ret != 0:
        raise TesseractError(ret, "unable to get languages")

    return [lang for lang in langs if lang and lang[-1] != ':']


def get_version():
    """
    Returns Tesseract version.

    Returns:
        A tuple corresponding to the version (for instance, (3, 0, 1) for 3.01)

    Exception:
        TesseractError --- Unable to run tesseract or to parse the version
    """
    _set_environment()

    command = [TESSERACT_CMD, "-v"]

    proc = subprocess.Popen(command,
                            startupinfo=g_subprocess_startup_info,
                            creationflags=g_creation_flags,
                            stdout=subprocess.PIPE,
                            stderr=subprocess.STDOUT)
    ver_string = proc.stdout.read()
    ver_string = ver_string.decode('utf-8')
    ret = proc.wait()
    if ret not in (0, 1):
        raise TesseractError(ret, ver_string)

    try:
        ver_string = ver_string.split(" ")[1]

        els = ver_string.split(".")
        els = [digits_only(x) for x in els]
        major = els[0]
        minor = els[1]
        upd = 0
        if len(els) >= 3:
            upd = els[2]
        version = (major, minor, upd)
        if version == (0, 0, 0):
            raise TesseractError(
                ret, ("Unable to parse Tesseract version (not a number): [%s]"
                      % (ver_string)))
        return version
    except IndexError:
        raise TesseractError(
            ret, ("Unable to parse Tesseract version (spliting failed): [%s]"
                  % (ver_string)))
